data/kernels/lut3d.cl

```diff
@@ -0,0 +1,225 @@
+/*
+    This file is part of darktable,
+    copyright (c) 2019 philippe weyland.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "common.h"
+
+kernel void
+lut3d_tetrahedral(read_only image2d_t in, write_only image2d_t out, const int width, const int height,
+           global float *clut, const int level)
+{
+  int4 rgbi = (int4)(0);
+  float4 rgbd = (float4)(0.0f);
+  const int level2 = level * level;
+  const int x = get_global_id(0);
+  const int y = get_global_id(1);
+
+  if(x >= width || y >= height) return;
+
+  float4 input = read_imagef(in, sampleri, (int2)(x, y));
+  float4 output = (float4)(0.0f);
+
+  input.x = (input.x < 0.0f) ? 0.0f : (input.x > 1.0f) ? 1.0f : input.x;
+  input.y = (input.y < 0.0f) ? 0.0f : (input.y > 1.0f) ? 1.0f : input.y;
+  input.z = (input.z < 0.0f) ? 0.0f : (input.z > 1.0f) ? 1.0f : input.z;
+
+  rgbd = input * (float)(level - 1); 
+  rgbi = min( max( convert_int4(rgbd), 0), (int)(level - 2));
+
+// delta r, g, b
+  rgbd = rgbd - convert_float4(rgbi); 
+
+// indexes of P000 to P111 in clut
+  const int color = rgbi.x + rgbi.y * level + rgbi.z * level2;
+  const int i000 = color * 3 ;  // P000
+  const int i100 = i000 + 3;  // P100
+  const int i010 = (int)(color + level) * 3;  // P010
+  const int i110 = i010 + 3;  //P110
+  const int i001 = (int)(color + level2) * 3;  // P001
+  const int i101 = i001 + 3;  // P101
+  const int i011 = (int)(color + level + level2) * 3;  // P011
+  const int i111 = i011 + 3;  // P111
+
+// float3 are 16 bytes aligned. So no way to set clut as an array of float3.
```

Comment:
`float3` does work. The problem is that a `float3` array has the same footprint than the equivalent float4 array. As the clut is an array of `float[3]` it cannot be read as an array of `float3`.
We could build the clut as an array of `float[4]` <=> `float4` at the expense of 33% increase of clut size if necessary.

Reply:
Ok, that's fine. You really don't want to increase the size of the buffers you send to the GPU if you don't have to, so that works. I was just curious.